Sources/TensorFlow/Layers/SequentialComposition.swift

```diff
@@ -0,0 +1,57 @@
+// Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// The sequential composition of the elements of some base collection of
+/// layers.
+///
+/// The first element of the base collection is applied first by the composite,
+/// so it is the inner call of the composition or the last in the sequence of
+/// composed layers in “f ∘ g” notation.
+public struct SequentialComposition<Base: Collection>
+  where Base.Element : Layer, Base.Element.Input == Base.Element.Output
```

Comment:
Currently, this patch makes it easy to apply a collection of `Layer`-conforming elements in sequence.

However, `SequentialComposition` is missing crucial conformances to `Differentiable` and `Layer`. `SequentialComposition.callAsFunction` should be `@differentiable(wrt: (self, input))`.

@dabrahams: would you like look into these changes?

Reply:
Here's a test case for the final desired functionality (code untested):
```swift
let (output, grad) =
  valueWithGradient(at: layers, input) { $0.sequentiallyComposed($1).sum() }
let (layersGrad, inputGrad) = grad

let (expectedOutput, expectedGrad) =
  valueWithGradient(at: input, layers[0], layers[1], layers[2]) { $0.sequenced(through: $1, $2, $3) }
let (expectedInputGrad, expectedLayer0Grad, expectedLayer1Grad, expectedLayer2Grad) = expectedGrad

XCTAssertEqual(output, expectedOutput)
XCTAssertEqual(inputGrad, expectedInputGrad)
XCTAssertEqual(layersGrad[0], expectedLayer0Grad)
XCTAssertEqual(layersGrad[1], expectedLayer1Grad)
XCTAssertEqual(layersGrad[2], expectedLayer2Grad)
```